controllers/state_of_the_world.go

```diff
@@ -184,29 +187,24 @@ func finalStepsWorkflow(client *dynamic.DynamicClient, isIstioInstalled, isEnvoy
 	return workflow
 }
 
-// GetOldestKuadrant returns the oldest kuadrant resource from a list of kuadrant resources that is not marked for deletion.
-func GetOldestKuadrant(kuadrants []*kuadrantv1beta1.Kuadrant) (*kuadrantv1beta1.Kuadrant, error) {
-	if len(kuadrants) == 1 {
-		return kuadrants[0], nil
-	}
-	if len(kuadrants) == 0 {
-		return nil, fmt.Errorf("empty list passed")
-	}
-	oldest := kuadrants[0]
-	for _, k := range kuadrants[1:] {
-		if k == nil || k.DeletionTimestamp != nil {
-			continue
-		}
-		if oldest == nil {
-			oldest = k
-			continue
-		}
-		if k.CreationTimestamp.Before(&oldest.CreationTimestamp) {
-			oldest = k
+var ErrNoKandrantResource = fmt.Errorf("no kuadrant resources in topology")
+
+// GetKuadrant returns the oldest Kuadrant from the root objects in the topology
+func GetKuadrant(topology *machinery.Topology) (*kuadrantv1beta1.Kuadrant, error) {
+	kuadrantList := lo.FilterMap(topology.Objects().Roots(), func(item machinery.Object, _ int) (controller.Object, bool) {
+		k, ok := item.(controller.Object)
+		if ok && k.GetObjectKind().GroupVersionKind().GroupKind() == kuadrantv1beta1.KuadrantGroupKind && k.GetDeletionTimestamp() == nil {
+			return k, true
 		}
+		return nil, false
+	})
+	if len(kuadrantList) == 1 {
+		return kuadrantList[0].(*kuadrantv1beta1.Kuadrant), nil
```

Comment:
This optimisation is probably not needed. If the number of elements in the slice is 1, then sorting is trivial.

Reply:
The way I see this is we have to do the zero length check, so if the length check has to be done we should short circuit sooner. With that said possible I should have added this to a switch statement and not use the if statements.

Reply:
The zero length is correct. I'm saying we don't need the check `if len(kuadrantList) == 1`.

Reply:
I am happy to remove it. To me it seemed more logical to test that the list is how we expect it to be 99% of the time and if not in that state then go do the sorting.